### PR TITLE
feat(storage, windows): Add windows support

### PR DESCRIPTION
**Pull request by cynthiajoan**
## Description

*Replace this paragraph with a description of what this PR is doing. If you're modifying existing behavior, describe the existing behavior, how this PR is changing it, and what motivated the change.*

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database](https://github.com/flutter/flutter/issues). Indicate, which of these issues are resolved or fixed by this PR. Note that you'll have to prefix the issue numbers with flutter/flutter#.*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,807 @@
...
+      return "download-size-exceeded";
+    case firebase::storage::kErrorCancelled:
+      return "canceled";
+
+    default:
+      return "unknown";
+  }
+}
+
+std::string FirebaseStoragePlugin::GetStorageErrorMessage(Error storageError) {
```
Comment (Lyokone) [inline packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp:164]: For the error message, the future is supposed to contain the error message directly
`completed_future.error_message();`
Maybe it's not the case with Storage ?

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,807 @@
...
+    events_ = std::move(events);
+
+    TaskStateListener putStringListener = TaskStateListener(events_.get());
+    StorageReference reference = storage_->GetReference(reference_path_);
+    Future<Metadata> future_result =
+        reference.PutFile(file_path_.c_str(), &putStringListener, controller_);
+
+    ::Sleep(1);
+    future_result.OnCompletion([this](const Future<Metadata>& data_result) {
+      if (data_result.error() == 0) {
```
Comment (Lyokone) [inline packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp:554]: In Firestore we had the 
`completed_future.error() == firebase::firestore::kErrorOk`
to check if we had no errors, I wonder if it's a bit more readable

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,807 @@
...
+      std::unique_ptr<flutter::EventSink<flutter::EncodableValue>>&& events)
+      override {
+    events_ = std::move(events);
+
+    TaskStateListener putStringListener = TaskStateListener(events_.get());
+    StorageReference reference = storage_->GetReference(reference_path_);
+    Future<Metadata> future_result =
+        reference.PutFile(file_path_.c_str(), &putStringListener, controller_);
+
+    ::Sleep(1);
```
Comment (Lyokone) [inline packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp:552]: Is this Sleep necessary? If so maybe we should add some comments to explain why

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,807 @@
...
+  Future<size_t> future_result = cpp_reference.GetBytes(byte_buffer, max_size);
+  ::Sleep(1);
+  future_result.OnCompletion(
+      [result, byte_buffer](const Future<size_t>& data_result) {
+        if (data_result.error() == 0) {
+          size_t vector_size = *data_result.result();
+          std::vector<uint8_t> vector_buffer(byte_buffer,
+                                             byte_buffer + vector_size);
+
+          // result(vector_buffer);
```
Comment (Lyokone) [inline packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp:404]: Is this meant to be commented?

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,807 @@
...
+    std::unique_lock<std::mutex> lock(mtx_);
+
+    TaskStateListener putStringListener = TaskStateListener(events_.get());
+    StorageReference reference = storage_->GetReference(reference_path_);
+    Future<size_t> future_result =
+        reference.GetFile(file_path_.c_str(), &putStringListener, controller_);
+
+    ::Sleep(1);
+    future_result.OnCompletion([this](const Future<size_t>& data_result) {
+      // TODO error handling
```
Comment (Lyokone) [inline packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp:624]: I think this TODO is done?

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp [hunk 6]
```diff
@@ -0,0 +1,807 @@
...
+    const PigeonStorageFirebaseApp& app,
+    const PigeonStorageReference& reference,
+    const PigeonSettableMetadata& metadata,
+    std::function<void(ErrorOr<PigeonFullMetaData> reply)> result) {
+  StorageReference cpp_reference =
+      GetCPPStorageReferenceFromPigeon(app, reference);
+  Metadata cpp_meta;
+  GetMetadataFromPigeon(metadata, &cpp_meta);
+  Future<Metadata> future_result = cpp_reference.UpdateMetadata(cpp_meta);
+  ::Sleep(1);
```
Comment (Lyokone) [inline packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp:756]: Same thing for the Sleep here, adding a comment would be nice

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp (hunk 4 above)
Comment (cynthiajoan) [reply packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp]: fixed

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp (hunk 2 above)
Comment (cynthiajoan) [reply packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp:560]: done

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp (hunk 5 above)
Comment (cynthiajoan) [reply packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp:625]: done

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp (hunk 1 above)
Comment (cynthiajoan) [reply packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp:164]: it is, however it is not exact same words with the other two platforms (iOS/Android), so do a conversion here for consistency.

File: packages/firebase_auth/firebase_auth/example/windows/flutter/CMakeLists.txt [hunk 7]
```diff
@@ -64,7 +64,7 @@ apply_standard_settings(flutter_wrapper_plugin)
 set_target_properties(flutter_wrapper_plugin PROPERTIES
   POSITION_INDEPENDENT_CODE ON)
 set_target_properties(flutter_wrapper_plugin PROPERTIES
-  CXX_VISIBILITY_PRESET hidden)
```
[the remark this answers is not in the document]
Comment (cynthiajoan) [reply packages/firebase_auth/firebase_auth/example/windows/flutter/CMakeLists.txt]: cleaned up

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp (hunk 3 above)
Comment (cynthiajoan) [reply packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp]: done, add the comments

File: packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp (hunk 6 above)
Comment (cynthiajoan) [reply packages/firebase_storage/firebase_storage/windows/firebase_storage_plugin.cpp]: added the comments

**Review by russellwheatley — APPROVED** [review]

**Review by Lyokone — APPROVED** [review]

**Pull request merged**